[volume-5] 인덱스 및 Redis 캐시 적용

junoade opened the pull request:
## 📌 Summary
<!--
    어떤 기능/이슈를 해결했는지 요약해주세요.
-->
- 상품목록 조회 서비스/API 개선 
- 대량 데이터 테스트를 위한 sql 코드 작성
- 상품목록 조회 성능 이슈 확인 및 분석
- 상품목록 조회에 대한 Redis 캐시 적용 (애노테이션 기반 스프링 AOP 활용)
- 전/후 API p95 성능지표 확인
- TTL 설정 스프링 빈 등록
- Page 결과 캐시 저장시 value 직렬화 방식 현재는 jdk 로. 추후 고민

## 💬 Review Points
<!--
    리뷰어가 더 효과적인 리뷰를 할 수 있도록 도와주는 부분입니다.
    (1) 리뷰어가 중점적으로 봐줬으면 하는 부분
    (2) 고민했던 설계 포인트나 로직
    (3) 리뷰어가 확인해줬으면 하는 테스트 케이스나 예외 상황
    (4) 기타 리뷰어가 참고해야 할 사항
-->

> 금주에는 이상한 회사 이슈(노조 투쟁) 및 그로 인한 감기 이슈가 있어서 과제에 시간을 많이 투자하지 못하였습니다 ㅠㅠ 주말에 보완하겠습니다.

1. 성능 이슈 확인 후 테이블 구조를 변경하고 이전 소스와 비교하는 ASIS/TOBE 테스트를 해보려고 하였는데 깃 별도의 브랜치를 만들고 테스트를 하면 비교가 원할할 것 같다는 생각이 들었습니다! 실무에서는 어떻게 하시는 지 궁금합니다.

2. 상품목록 조회 API 에서 필터들을 여러개 걸면 
- 브랜드ID 전달시 성능 OK
- 브랜드ID 전달 안할 경우 성능 BAD
이런식으로 필터의 유무에 따라 쿼리 성능에도 좀 영향이 있는 것 같았습니다.
동적쿼리 성능에 대한 고민일 것 같은데..  하나의 API에 권장되는 필터 개수 이런게 있을까요?


## ✅ Checklist
<!--
    해당 작업이 완료되었는지 확인하기 위한 체크리스트입니다.
    리뷰어가 확인해야 할 사항을 포함합니다.
    계획중이나 아직 완료되지 않은 작업 또한 `TODO -` 로 작성해주세요.  

    ex.
    - [ ] 테스트 코드 포함
    - [ ] 불필요한 코드 제거
    - [ ] README or 주석 보강 (필요 시)
-->
### Index
- [x]  상품 목록 API에서 brandId 기반 검색, 좋아요 순 정렬 등을 처리했다
- [x]  조회 필터, 정렬 조건별 유즈케이스를 분석하여 인덱스를 적용하고 전 후 성능비교를 진행했다

### ❤️ Structure
- [x]  상품 목록/상세 조회 시 좋아요 수를 조회 및 좋아요 순 정렬이 가능하도록 구조 개선을 진행했다
- [x]  좋아요 적용/해제 진행 시 상품 좋아요 수 또한 정상적으로 동기화되도록 진행하였다

### ⚡ Cache
- [x]  Redis 캐시를 적용하고 TTL 또는 무효화 전략을 적용했다
- [x]  캐시 미스 상황에서도 서비스가 정상 동작하도록 처리했다.

## 📎 References
<!--
  (Optional: 참고 자료가 없는 작업 - 단순 버그 픽스 등 의 경우엔 해당 란을 제거해주세요 !)
  리뷰어가 참고할 수 있는 추가적인 정보나 문서, 링크 등을 작성해주세요.
  예시:
  - 관련 문서 링크
  - 관련 정책 링크
-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## 릴리스 노트

* **New Features**
  * 상품 목록 조회 API 추가 (브랜드 필터링, 정렬 옵션 지원)
  * 상품 상세 조회 API 추가
  * 페이지네이션 및 좋아요 수 집계 기능 제공

* **Tests**
  * 상품 조회 서비스 테스트 추가

* **Chores**
  * 대규모 테스트 데이터 초기화 지원 추가

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

junoade committed:
feat: 상품 목록 조회 서비스 개선

junoade committed:
feat: 상품 목록 조회 서비스 테스트

junoade committed:
feat: 상품 목록 조회 API 개발 완료

junoade committed:
feat: 상품 상세조회 API 개발 완료

junoade committed:
feat: 상품 가격 정렬 추가

junoade committed:
test: 대량데이터 테스트

junoade committed:
test: 대량데이터 테스트 수정

junoade committed:
Merge pull request #25 from junoade/week5-feature-productList

Week5 feature product list

looper88888 commented:
🔖 Index
[] 상품 목록 API에서 brandId 기반 검색, 좋아요 순 정렬 등을 처리했다
[] 조회 필터, 정렬 조건별 유즈케이스를 분석하여 인덱스를 적용하고 전 후 성능비교를 진행했다
❤️ Structure
[] 상품 목록/상세 조회 시 좋아요 수를 조회 및 좋아요 순 정렬이 가능하도록 구조 개선을 진행했다
[] 좋아요 적용/해제 진행 시 상품 좋아요 수 또한 정상적으로 동기화되도록 진행하였다
⚡ Cache
[] Redis 캐시를 적용하고 TTL 또는 무효화 전략을 적용했다
[] 캐시 미스 상황에서도 서비스가 정상 동작하도록 처리했다.

@coderabbitai 위 체크리스트가 맞는지 확인해줘.

junoade committed:
feat: 상품좋아요 집계 반정규화 / 상품정보 조회 QueryDsl 추상화

junoade committed:
feat: ddl-auto : none

junoade committed:
fix: likeCount NPE 문제 일시 해소

junoade committed:
feat: 반정규화 후 Product 테이블 조회 쿼리 테스트

- ddl-auto:none 후 db 스키마 변경 및 상품좋아요 백필 수행

```sql
ALTER TABLE product ADD COLUMN like_count BIGINT NOT NULL DEFAULT 0;

UPDATE product p
JOIN (
  SELECT product_id, COUNT(*) AS cnt
  FROM product_like
  GROUP BY product_id
) pl ON pl.product_id = p.id
SET p.like_count = pl.cnt;
```

junoade committed:
fix: unimport unusage class

junoade committed:
fix: 반정규화에 따른 멱등한 좋아요 등록/삭제 대응

junoade committed:
Merge pull request #26 from junoade/week5-refactor-denormalization

Week5 refactor denormalization

junoade committed:
feat: 상품목록 조회 캐싱 적용

junoade committed:
feat: 상품목록 조회 TTL 적용

junoade committed:
Merge pull request #27 from junoade/week5-cache

Week5 cache
